src/main/java/org/springframework/data/jpa/mapping/JpaMetamodelMappingContext.java

```diff
@@ -109,6 +112,11 @@ public boolean hasPersistentEntityFor(Class<?> type) {
 		return super.hasPersistentEntityFor(type) || models.isMetamodelManagedType(type);
 	}
 
+	@EventListener
```

Comment:
Spring library components are not supposed to use `@EventListener` (it's a choice for application developers, but it's really inefficient, so not a good idea for libraries). You could put this in a `@PreDestroy` though, I expect (or `DisposableBean` probably better).

Reply:
Thanks for the review! I just updated my PR to use a `DisposableBean` instead.

Reply:
Well, implementing `DisposableBean` in class `JpaMetamodelMappingContext` was not triggering anything when application context is closed. Since `JpaMetamodelMappingContextFactoryBean` already implements `DisposableBean`, I decided to update my PR to clean the cache here instead.
This time, the cache is cleaned when applicationContext is closed.